File: C++/Day03.cpp
#include "AH.h"

namespace Day03
{

	std::pair<int, int> fn(const std::vector<std::string> ss) {
		int part1 = 0;
		int part2 = 0;
		std::map<int, int> gears;
		for (int r = 0; r < ss.size(); ++r) {
			const auto s = ss[r];
			for (int c = 0; c < s.size(); ++c) {
				const auto ch = s.at(c);
				if (isdigit(ch)) {
					auto idx = s.find_first_not_of("0123456789", c);
					if (idx == std::string::npos) // the number runs right to the edge of the grid
						idx = s.size();
					// does string have a neighbouring symbol?
					bool is_good = false;
					int gear_idx = 0;
					const auto digits = s.substr(c, idx - c);
					for (int ir = std::max(0, r - 1); ir < std::min((int)ss.size(), r + 2); ++ir) {
						for (int ic = std::max(0, c - 1); ic < std::min((int)s.size(), (int)idx + 1); ++ic) {
							const auto qs = ss[ir];
							if ((qs.at(ic) != '.') && (!isdigit(qs.at(ic)))) {
								is_good = true;
								if (qs.at(ic) == '*') {
									gear_idx = ir * s.size() + ic;
								}
								break;
							}
						}
						if (is_good) break;
					}

					if (is_good) {
						const auto v = stoi(s.substr(c, idx - c));
						part1 += v;
						if (gear_idx != 0) {
							if (gears.count(gear_idx)) // we have seen this gear before
								part2 += gears[gear_idx] * v;
							else
								gears[gear_idx] += v;
						}
					}

					c = idx;
				}
			}
		}

		return std::make_pair(part1, part2);
	}

	int Run(const std::string& filename)
	{
		const std::vector<std::string> inputLines = AH::ReadTextFile(filename);

		auto [part1, part2] = fn(inputLines);

		AH::PrintSoln(3, part1, part2);

		return 0;
	}

}

File: Go/Day03.go
package main

import (
	"strconv"
	"unicode"

	AH "./adventhelper"
)

func fn(ss []string) (part1 int, part2 int) {
	part1, part2 = 0, 0

	gears := make(map[int]int)

	for r := 0; r < len(ss); r++ {
		s := ss[r]
		for c := 0; c < len(s); c++ {
			digits := AH.TakeWhileDigit(s[c:])
			n := len(digits)
			if len(digits) > 0 { // we have a number
				hasNbr := false
				gearIdx := 0
				for ir := AH.Max(0, r-1); ir < AH.Min(len(ss), r+2); ir++ {
					for ic := AH.Max(0, c-1); ic < AH.Min(len(s), c+n+1); ic++ {
						qs := ss[ir]
						qc := AH.RuneAt(qs, ic)
						if (qc != '.') && (!unicode.IsDigit(qc)) {
							hasNbr = true
							if qc == '*' {
								gearIdx = ir*len(s) + ic
							}
							break
						}
					}
					if hasNbr {
						break
					}
				}

				if hasNbr {
					v, _ := strconv.Atoi(digits)
					part1 += v
					if gearIdx != 0 {
						_, ok := gears[gearIdx]
						if ok {
							part2 += gears[gearIdx] * v
						} else {
							gears[gearIdx] = v
						}

					}
				}

				c += len(digits)
			}
		}
	}

	return
}

func main() {
	inputLines, _ := AH.ReadStrFile("../input/input03.txt")

	part1, part2 := fn(inputLines)

	AH.PrintSoln(3, part1, part2)

	return
}
